File: templates/production-adapters/opencode-http-session/typescript/runner.ts
import { mkdir, readFile, writeFile } from "node:fs/promises"
import { join, resolve } from "node:path"

import type { GateContract, WorkflowBundle } from "./contracts"
import { loadBundleContract } from "./contracts"
import { appendJsonLine, readJsonObject } from "./json-store"
import { OpencodeHttpClient, type OpencodeHttpConfig, OpencodeHttpError } from "./opencode-http"
import { ensureServer, stopServer, writeServerState } from "./opencode-server"
import { clearStageOutputs, ensureStageOutputs, renderPrompt, requiredOutputPaths } from "./prompt-assembly"
import { parseRunnerArgs } from "./runner-cli"
import { prepareRunDirectory, resolveRunDirectory } from "./runner-inputs"
import { type SessionRecord, SessionStore } from "./session-store"

export const WORKFLOW_ROOT = import.meta.dir
export const REPO_ROOT = resolve(WORKFLOW_ROOT, "..", "..")
export const DEFAULT_WORKFLOW = join(WORKFLOW_ROOT, "workflow.yml")
export const DEFAULT_OPENCODE_URL = "http://127.0.0.1:5096"

const MAX_REVIEW_ATTEMPTS = 2
const STAGE_POLL_INTERVAL_MS = 250
const STAGE_HARD_TIMEOUT_MULTIPLIER = 4
const OPEN_ASSISTANT_MAX_WAIT_MS = 3_600_000
const ATTACH_GUIDE_FILENAME = "opencode-attach-commands.txt"

export async function main(): Promise<number> {
  const parsed = parseRunnerArgs(Bun.argv.slice(2), { opencodeUrl: DEFAULT_OPENCODE_URL, workflowPath: DEFAULT_WORKFLOW })
  const bundle = await loadBundleContract(parsed.workflowPath)
  const runRoot = resolve(REPO_ROOT, bundle.runtime.runDirectory.root)
  const runDirectory = await resolveRunDirectory({
    deriveFromInput: bundle.runtime.runDirectory.deriveFromInput,
    explicitRunDirectory: parsed.runDirectory,
    inputArtifact: bundle.manifest.run.inputArtifact,
    repoRoot: REPO_ROOT,
    runRoot,
    workflowRoot: bundle.workflowRoot,
  })
  await prepareRunDirectory({
    freshRun: parsed.freshRun,
    markerName: bundle.runtime.runDirectory.freshRunMarker,
    runDirectory,
    runRoot,
    seedArtifacts: bundle.manifest.run.seedArtifacts,
    workflowRoot: bundle.workflowRoot,
  })
  if (parsed.stopAfterSeed) {
    return 0
  }
  const store = new SessionStore(runDirectory)
  const httpConfig: OpencodeHttpConfig = {
    baseUrl: parsed.opencodeUrl,
    password: process.env.OPENCODE_SERVER_PASSWORD,
    timeoutMs: parsed.agentTimeoutMs,
    username: process.env.OPENCODE_SERVER_USERNAME,
  }
  const server = await ensureServer({
    baseUrl: parsed.opencodeUrl,
    host: parsed.opencodeHost,
    noStart: parsed.noStartOpencode,
    password: httpConfig.password,
    port: parsed.opencodePort,
    runDirectory,
    timeoutMs: parsed.agentTimeoutMs,
    username: httpConfig.username,
    workdir: REPO_ROOT,
  })
  await writeServerState(join(runDirectory, "artifacts", "opencode-server.json"), server)
  const client = new OpencodeHttpClient({ ...httpConfig, baseUrl: server.baseUrl })
  await ensureWorkflowSessions({ bundle, client, runDirectory, store })
  await printObserverSummary({ bundle, runDirectory, serverUrl: server.baseUrl, store })
  try {
    return await runWorkflow({ agentTimeoutMs: parsed.agentTimeoutMs, bundle, client, runDirectory, serverUrl: server.baseUrl, store })
  } finally {
    stopServer(server)
  }
}

export async function runWorkflow(input: {
  readonly agentTimeoutMs: number
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly runDirectory: string
  readonly serverUrl: string
  readonly store: SessionStore
}): Promise<number> {
  let current = input.bundle.entry
  let reviewAttempts = 0
  for (let transition = 0; transition < 256; transition += 1) {
    const stage = input.bundle.manifest.stages[current]
    if (stage !== undefined) {
      await runStage({ ...input, stage: current })
      let nextNode = stage.routing.onSuccess
      if (await stageControlsDecision(input.runDirectory, current, input.bundle)) {
        const approved = await decisionMatchesContract(input.runDirectory, input.bundle)
        nextNode = approved ? stage.routing.onSuccess : stage.routing.onFailure
        if (!approved) {
          reviewAttempts += 1
          if (reviewAttempts > MAX_REVIEW_ATTEMPTS) {
            await input.store.setRuntimeState({ reason: "review decision remained unapproved after max attempts", stage: current, status: "failed" })
            return 1
          }
        }
      }
      if (nextNode === undefined) {
        await input.store.setRuntimeState({ stage: current, status: "completed" })
        return 0
      }
      current = nextNode
      continue
    }
    const gate = input.bundle.manifest.gates[current]
    if (gate !== undefined) {
      if (current === "finalize") {
        const result = await runGate({ bundle: input.bundle, gate: current, runDirectory: input.runDirectory })
        await input.store.setRuntimeState({ stage: current, status: result.ok ? "completed" : "failed" })
        return result.exitCode
      }
      const result = await retryGateWithFeedback({ ...input, gate: current })
      if (result.ok) {
        current = gate.pass ?? ""
        if (current.length === 0) {
          await input.store.setRuntimeState({ stage: result.label, status: "completed" })
          return result.exitCode
        }
        continue
      }
      if ((gate.maxAttempts ?? 1) > 1 && gate.fail?.endsWith("_feedback")) {
        await input.store.setRuntimeState({ stage: result.label, status: "failed" })
        return result.exitCode
      }
      current = gate.fail ?? ""
      if (current.length === 0) {
        await input.store.setRuntimeState({ stage: result.label, status: "failed" })
        return result.exitCode
      }
      continue
    }
    throw new Error(`workflow routed to unknown node: ${current}`)
  }
  await input.store.setRuntimeState({ reason: "workflow exceeded 256 node transitions", stage: current, status: "failed" })
  return 1
}

export async function retryGateWithFeedback(input: {
  readonly agentTimeoutMs: number
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly gate: string
  readonly runDirectory: string
  readonly serverUrl: string
  readonly store: SessionStore
}): Promise<{ readonly exitCode: number; readonly label: string; readonly ok: boolean }> {
  const gate = input.bundle.manifest.gates[input.gate]
  let last = { exitCode: 1, label: input.gate, ok: false }
  for (let attempt = 1; attempt <= (gate.maxAttempts ?? 1); attempt += 1) {
    last = await runGate({ bundle: input.bundle, gate: input.gate, runDirectory: input.runDirectory })
    if (last.ok) {
      return last
    }
    if (attempt < (gate.maxAttempts ?? 1) && gate.fail !== undefined && input.bundle.manifest.stages[gate.fail] !== undefined) {
      await runStage({ ...input, stage: gate.fail })
    }
  }
  return last
}

export async function runStage(input: {
  readonly agentTimeoutMs: number
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly runDirectory: string
  readonly serverUrl: string
  readonly stage: string
  readonly store: SessionStore
}): Promise<void> {
  const stageContract = input.bundle.manifest.stages[input.stage]
  if (stageContract === undefined) {
    throw new Error(`unknown stage: ${input.stage}`)
  }
  const session = await getOrCreateSession({ bundle: input.bundle, client: input.client, stage: stageContract.session, store: input.store })
  const attempt = session.promptCount + 1
  await clearStageOutputs({ bundle: input.bundle, repoRoot: REPO_ROOT, runDirectory: input.runDirectory, stage: input.stage })
  const rendered = await renderPrompt({ attempt, bundle: input.bundle, repoRoot: REPO_ROOT, runDirectory: input.runDirectory, stage: input.stage })
  await logRunnerEvent(input.runDirectory, input.stage, `submitting prompt ${rendered.promptPath}`)
  await input.client.sendPromptAsync(session.sessionId, { agent: session.agent, prompt: rendered.text })
  await input.store.upsert(stageContract.session, { agent: session.agent, promptCount: attempt, sessionId: session.sessionId })
  await input.store.appendTurn({ agent: session.agent, attempt, kind: "stage-prompt", promptPath: rendered.promptPath, session: stageContract.session, sessionId: session.sessionId, stage: input.stage })
  await waitForStageCompletion({ agentTimeoutMs: input.agentTimeoutMs, bundle: input.bundle, client: input.client, runDirectory: input.runDirectory, sessionId: session.sessionId, stage: input.stage })
  await ensureStageOutputs({ bundle: input.bundle, repoRoot: REPO_ROOT, runDirectory: input.runDirectory, stage: input.stage })
  await input.store.setRuntimeState({ serverUrl: input.serverUrl, stage: input.stage, status: "completed" })
}

export async function runGate(input: {
  readonly bundle: WorkflowBundle
  readonly gate: string
  readonly runDirectory: string
}): Promise<{ readonly exitCode: number; readonly label: string; readonly ok: boolean }> {
  const gate = input.bundle.manifest.gates[input.gate]
  if (gate === undefined) {
    throw new Error(`unknown gate: ${input.gate}`)
  }
  const command = renderCommand(gate, input.runDirectory)
  const subprocess = Bun.spawn(command, {
    cwd: gateCwd(input.bundle, input.gate),
    stderr: "pipe",
    stdout: "pipe",
  })
  const [stdout, stderr, exitCode] = await Promise.all([
    new Response(subprocess.stdout).text(),
    new Response(subprocess.stderr).text(),
    subprocess.exited,
  ])
  await mkdir(join(input.runDirectory, "artifacts", "logs"), { recursive: true })
  await writeFile(join(input.runDirectory, "artifacts", "logs", `${input.gate}.stdout.txt`), stdout)
  await writeFile(join(input.runDirectory, "artifacts", "logs", `${input.gate}.stderr.txt`), stderr)
  await appendJsonLine(join(input.runDirectory, "artifacts", "events.jsonl"), { exitCode, gate: input.gate, kind: "gate", ok: exitCode === 0 })
  return { exitCode, label: input.gate, ok: exitCode === 0 }
}

export async function getOrCreateSession(input: {
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly stage: string
  readonly store: SessionStore
}): Promise<SessionRecord> {
  const sessionContract = input.bundle.manifest.sessions[input.stage]
  if (sessionContract === undefined) {
    throw new Error(`session contract missing for ${input.stage}`)
  }
  const existing = await input.store.get(input.stage)
  if (existing !== undefined && (await input.client.getSession(existing.sessionId)) !== undefined) {
    return existing
  }
  const created = await input.client.createSession(`${input.bundle.manifest.run.workflowName}:${input.stage}`)
  return input.store.upsert(input.stage, { agent: sessionContract.agent, promptCount: 0, sessionId: created.sessionId })
}

export async function ensureWorkflowSessions(input: {
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly runDirectory: string
  readonly store: SessionStore
}): Promise<void> {
  for (const stage of Object.keys(input.bundle.manifest.sessions).sort()) {
    await getOrCreateSession({ bundle: input.bundle, client: input.client, stage, store: input.store })
  }
  await refreshAttachGuide({ ...input, serverUrl: "<pending>" })
}

export async function waitForStageCompletion(input: {
  readonly agentTimeoutMs: number
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly runDirectory: string
  readonly sessionId: string
  readonly stage: string
}): Promise<void> {
  const deadline = Date.now() + Math.max(input.agentTimeoutMs * STAGE_HARD_TIMEOUT_MULTIPLIER, OPEN_ASSISTANT_MAX_WAIT_MS)
  while (Date.now() < deadline) {
    const observation = await observeStageCompletion(input)
    await appendJsonLine(join(input.runDirectory, "artifacts", "events.jsonl"), { kind: "stage-observe", stage: input.stage, ...observation, outputFingerprint: undefined })
    if (observation.assistantCompleted && observation.outputsComplete) {
      return
    }
    await Bun.sleep(STAGE_POLL_INTERVAL_MS)
  }
  throw new Error(`stage ${input.stage} did not complete before timeout`)
}

export async function observeStageCompletion(input: {
  readonly bundle: WorkflowBundle
  readonly client: OpencodeHttpClient
  readonly runDirectory: string
  readonly sessionId: string
  readonly stage: string
}): Promise<{
  readonly assistantCompleted: boolean
  readonly assistantStarted: boolean
  readonly messagesSupported: boolean
  readonly missingOutputs: readonly string[]
  readonly openAssistant: boolean
  readonly outputFingerprint: readonly [string, boolean, number | undefined][]
  readonly outputsComplete: boolean
}> {
  let messagesSupported = true
  let assistantStarted = false
  let assistantCompleted = false
  try {
    const messages = await input.client.listMessages(input.sessionId)
    const assistantMessages = messages.filter((message) => message.role === "assistant")
    assistantStarted = assistantMessages.length > 0
    assistantCompleted = assistantMessages.some(messageLooksComplete)
  } catch (error) {
    if (!(error instanceof OpencodeHttpError)) {
      throw error
    }
    messagesSupported = false
    assistantCompleted = true
  }
  const outputFingerprint: [string, boolean, number | undefined][] = []
  const missingOutputs: string[] = []
  for (const path of requiredOutputPaths({ bundle: input.bundle, repoRoot: REPO_ROOT, runDirectory: input.runDirectory, stage: input.stage })) {
    const file = Bun.file(path)
    const exists = await file.exists()
    outputFingerprint.push([path, exists, exists ? file.size : undefined])
    if (!exists) {
      missingOutputs.push(path)
    }
  }
  return {
    assistantCompleted,
    assistantStarted,
    messagesSupported,
    missingOutputs,
    openAssistant: assistantStarted && !assistantCompleted,
    outputFingerprint,
    outputsComplete: missingOutputs.length === 0,
  }
}

export async function printObserverSummary(input: {
  readonly bundle: WorkflowBundle
  readonly runDirectory: string
  readonly serverUrl: string
  readonly store: SessionStore
}): Promise<void> {
  await refreshAttachGuide(input)
  const guidePath = join(input.runDirectory, "artifacts", ATTACH_GUIDE_FILENAME)
  console.log(`[runner] runDirectory=${input.runDirectory}`)
  console.log(`[runner] opencodeServer=${input.serverUrl}`)
  console.log(`[runner] attachGuide=${guidePath}`)
  console.log("[runner] attach commands:")
  console.log(await readFile(guidePath, "utf8"))
}

export async function refreshAttachGuide(input: {
  readonly bundle: WorkflowBundle
  readonly runDirectory: string
  readonly serverUrl: string
  readonly store: SessionStore
}): Promise<void> {
  const lines = [
    "# OpenCode attach commands",
    `server=${input.serverUrl}`,
    `runDirectory=${input.runDirectory}`,
    "",
    "# Use /exit to leave the TUI client. It should not cancel server-owned execution.",
  ]
  const sessions = await input.store.sessions()
  for (const stage of Object.keys(input.bundle.manifest.sessions).sort()) {
    const record = sessions[stage]
    if (record === undefined) {
      continue
    }
    lines.push("", `# ${stage}`, attachCommand(input.serverUrl, record.id))
  }
  await writeFile(join(input.runDirectory, "artifacts", ATTACH_GUIDE_FILENAME), `${lines.join("\n")}\n`)
}

export function attachCommand(serverUrl: string, sessionId: string): string {
  return ["opencode", "attach", shellQuote(serverUrl), "--dir", shellQuote(REPO_ROOT), "--session", shellQuote(sessionId)].join(" ")
}

function renderCommand(gate: GateContract, runDirectory: string): string[] {
  return gate.command.map((part) => part.replaceAll("{{runDirectory}}", runDirectory))
}

function gateCwd(bundle: WorkflowBundle, gate: string): string {
  const cwd = bundle.gateCwds[gate]
  return cwd === undefined ? bundle.workflowRoot : resolve(bundle.workflowRoot, cwd)
}

async function stageControlsDecision(runDirectory: string, stage: string, bundle: WorkflowBundle): Promise<boolean> {
  if (stage === "review") {
    return true
  }
  const stageContract = bundle.manifest.stages[stage]
  return stageContract.requiredOutputs.includes(bundle.manifest.decision.artifact) && (await Bun.file(join(runDirectory, bundle.manifest.decision.artifact)).exists())
}

async function decisionMatchesContract(runDirectory: string, bundle: WorkflowBundle): Promise<boolean> {
  const path = join(runDirectory, bundle.manifest.decision.artifact)
  if (!(await Bun.file(path).exists())) {
    return false
  }
  let cursor: unknown = await readJsonObject(path)
  for (const segment of bundle.manifest.decision.jsonPath) {
    if (typeof cursor !== "object" || cursor === null || Array.isArray(cursor) || !(segment in cursor)) {
      return false
    }
    cursor = (cursor as Record<string, unknown>)[segment]
  }
  return cursor === bundle.manifest.decision.equals
}

async function logRunnerEvent(runDirectory: string, label: string, message: string): Promise<void> {
  await appendJsonLine(join(runDirectory, "artifacts", "runner-events.jsonl"), { label, message, time: Date.now() / 1000 })
}

function messageLooksComplete(message: Record<string, unknown>): boolean {
  const status = message.status
  if (typeof status === "string" && ["completed", "done", "idle"].includes(status.toLowerCase())) {
    return true
  }
  if (Array.isArray(message.parts) || typeof message.content === "string") {
    return true
  }
  return false
}

function shellQuote(value: string): string {
  return `'${value.replaceAll("'", "'\\''")}'`
}

if (import.meta.main) {
  process.exit(await main())
}

File: templates/production-adapters/opencode-http-session/python/runner.py
from __future__ import annotations

import hashlib
import json
import os
import shlex
import subprocess
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Final

from contracts import GateContract, WorkflowBundleContract, load_bundle_contract
from opencode_http import OpencodeHttpClient, OpencodeHttpConfig, OpencodeHttpError
from opencode_server import ensure_server, stop_server, write_server_state
from prompt_assembly import clear_stage_outputs, ensure_stage_outputs, render_prompt
from runner_cli import parse_args, resolve_runner_args
from runner_inputs import prepare_run_directory
from session_store import SessionRecord, SessionStore


WORKFLOW_ROOT: Final = Path(__file__).resolve().parent
REPO_ROOT: Final = WORKFLOW_ROOT.parents[1]
DEFAULT_WORKFLOW: Final = WORKFLOW_ROOT / "workflow.yml"
DEFAULT_OPENCODE_URL: Final = "http://127.0.0.1:5096"
MAX_REVIEW_ATTEMPTS: Final = 2
STAGE_POLL_INTERVAL_SEC: Final = 0.25
STAGE_HARD_TIMEOUT_MULTIPLIER: Final = 4.0
OPEN_ASSISTANT_MAX_WAIT_SEC: Final = 3600.0
ATTACH_GUIDE_FILENAME: Final = "opencode-attach-commands.txt"


@dataclass(frozen=True, slots=True)
class GateResult:
    label: str
    exit_code: int

    @property
    def ok(self) -> bool:
        return self.exit_code == 0


@dataclass(frozen=True, slots=True)
class StageWaitObservation:
    messages_supported: bool
    assistant_started: bool
    assistant_completed: bool
    open_assistant: bool
    outputs_complete: bool
    message_fingerprint: str | None
    output_fingerprint: tuple[tuple[str, bool, int | None], ...]
    missing_outputs: tuple[str, ...]


def main() -> int:
    parsed = parse_args(workflow_default=DEFAULT_WORKFLOW, opencode_url_default=DEFAULT_OPENCODE_URL)
    contract = load_bundle_contract(parsed.workflow_path)
    run_root = (REPO_ROOT / contract.runtime.run_directory.root).resolve()
    args = resolve_runner_args(
        parsed,
        repo_root=REPO_ROOT,
        run_root=run_root,
        derive_from_input=contract.runtime.run_directory.derive_from_input,
        workflow_root=contract.workflow_root,
        input_artifact=contract.manifest.run.input_artifact,
    )
    prepare_run_directory(
        run_dir=args.run_dir,
        run_root=run_root,
        workflow_root=contract.workflow_root,
        seed_paths=tuple(Path(path) for path in contract.manifest.run.seed_artifacts),
        marker_name=contract.runtime.run_directory.fresh_run_marker,
        fresh_run=args.fresh_run,
    )
    if args.stop_after_seed:
        return 0
    store = SessionStore(args.run_dir)
    server = ensure_server(
        base_url=args.opencode_url,
        host=args.opencode_host,
        port=args.opencode_port,
        no_start=args.no_start_opencode,
        timeout_sec=args.agent_timeout_sec,
        run_dir=args.run_dir,
        workdir=REPO_ROOT,
        username=os.environ.get("OPENCODE_SERVER_USERNAME"),
        password=os.environ.get("OPENCODE_SERVER_PASSWORD"),
    )
    write_server_state(args.run_dir / "artifacts" / "opencode-server.json", server)
    client = OpencodeHttpClient(
        OpencodeHttpConfig(
            server.base_url,
            args.agent_timeout_sec,
            os.environ.get("OPENCODE_SERVER_USERNAME"),
            os.environ.get("OPENCODE_SERVER_PASSWORD"),
        )
    )
    ensure_workflow_sessions(run_dir=args.run_dir, contract=contract, store=store, client=client)
    print_observer_summary(run_dir=args.run_dir, server_url=server.base_url, contract=contract, store=store)
    try:
        return run_workflow(
            run_dir=args.run_dir,
            server_url=server.base_url,
            agent_timeout_sec=args.agent_timeout_sec,
            contract=contract,
            store=store,
            client=client,
        )
    finally:
        stop_server(server)


def run_workflow(
    *,
    run_dir: Path,
    server_url: str,
    agent_timeout_sec: float,
    contract: WorkflowBundleContract,
    store: SessionStore,
    client: OpencodeHttpClient,
) -> int:
    current = contract.entry
    review_attempts = 0
    for _ in range(256):
        if current in contract.manifest.stages:
            stage_contract = contract.manifest.stages[current]
            run_stage(
                run_dir=run_dir,
                stage=current,
                server_url=server_url,
                agent_timeout_sec=agent_timeout_sec,
                contract=contract,
                store=store,
                client=client,
            )
            next_node = stage_contract.routing.on_success
            if _stage_controls_decision(run_dir=run_dir, stage=current, contract=contract):
                approved = decision_matches_contract(run_dir=run_dir, contract=contract)
                next_node = stage_contract.routing.on_success if approved else stage_contract.routing.on_failure
                if not approved:
                    review_attempts += 1
                    if review_attempts > MAX_REVIEW_ATTEMPTS:
                        store.set_runtime_state(
                            {
                                "stage": current,
                                "status": "failed",
                                "reason": "review decision remained unapproved after max attempts",
                            }
                        )
                        return 1
            if next_node is None:
                store.set_runtime_state({"stage": current, "status": "completed"})
                return 0
            current = next_node
            continue
        if current in contract.manifest.gates:
            gate = contract.manifest.gates[current]
            if current == "finalize":
                result = run_gate(run_dir=run_dir, gate=current, contract=contract)
                store.set_runtime_state({"stage": current, "status": "completed" if result.ok else "failed"})
                return result.exit_code
            result = retry_gate_with_feedback(
                run_dir=run_dir,
                gate=current,
                server_url=server_url,
                agent_timeout_sec=agent_timeout_sec,
                contract=contract,
                store=store,
                client=client,
            )
            if result.ok:
                current = gate.pass_node
                if current is None:
                    store.set_runtime_state({"stage": result.label, "status": "completed"})
                    return result.exit_code
                continue
            if gate.max_attempts > 1 and gate.fail_node is not None and gate.fail_node.endswith("_feedback"):
                store.set_runtime_state({"stage": result.label, "status": "failed"})
                return result.exit_code
            current = gate.fail_node
            if current is None:
                store.set_runtime_state({"stage": result.label, "status": "completed" if result.ok else "failed"})
                return result.exit_code
            continue
        raise RuntimeError(f"workflow routed to unknown node: {current}")
    store.set_runtime_state({"stage": current, "status": "failed", "reason": "workflow exceeded 256 node transitions"})
    return 1


def retry_gate_with_feedback(
    *,
    run_dir: Path,
    gate: str,
    server_url: str,
    agent_timeout_sec: float,
    contract: WorkflowBundleContract,
    store: SessionStore,
    client: OpencodeHttpClient,
) -> GateResult:
    gate_contract = contract.manifest.gates[gate]
    last = GateResult(gate, 1)
    for attempt in range(1, gate_contract.max_attempts + 1):
        last = run_gate(run_dir=run_dir, gate=gate, contract=contract)
        if last.ok:
            return last
        if attempt < gate_contract.max_attempts and gate_contract.fail_node in contract.manifest.stages:
            run_stage(
                run_dir=run_dir,
                stage=gate_contract.fail_node,
                server_url=server_url,
                agent_timeout_sec=agent_timeout_sec,
                contract=contract,
                store=store,
                client=client,
            )
    return last


def run_stage(
    *,
    run_dir: Path,
    stage: str,
    server_url: str,
    agent_timeout_sec: float,
    contract: WorkflowBundleContract,
    store: SessionStore,
    client: OpencodeHttpClient,
) -> None:
    stage_contract = contract.manifest.stages[stage]
    session = get_or_create_session(contract=contract, store=store, client=client, stage=stage_contract.session)
    attempt = session.prompt_count + 1
    clear_stage_outputs(
        stage,
        run_dir=run_dir,
        repo_root=REPO_ROOT,
        workflow_root=contract.workflow_root,
        manifest=contract.manifest,
    )
    rendered = render_prompt(
        stage=stage,
        workflow_root=contract.workflow_root,
        run_dir=run_dir,
        attempt=attempt,
        repo_root=REPO_ROOT,
        manifest=contract.manifest,
        input_artifact=contract.manifest.run.input_artifact,
    )
    log_runner_event(run_dir, stage, f"submitting prompt {rendered.prompt_path}")
    client.send_prompt_async(session.session_id, agent=session.agent, prompt=rendered.text)
    store.upsert(stage_contract.session, session_id=session.session_id, agent=session.agent, prompt_count=attempt)
    store.append_turn(
        {
            "kind": "stage-prompt",
            "stage": stage,
            "session": stage_contract.session,
            "sessionId": session.session_id,
            "agent": session.agent,
            "attempt": attempt,
            "promptPath": str(rendered.prompt_path),
        }
    )
    wait_for_stage_completion(
        run_dir=run_dir,
        stage=stage,
        session_id=session.session_id,
        client=client,
        agent_timeout_sec=agent_timeout_sec,
        contract=contract,
    )
    ensure_stage_outputs(
        stage,
        run_dir=run_dir,
        repo_root=REPO_ROOT,
        workflow_root=contract.workflow_root,
        manifest=contract.manifest,
    )
    store.set_runtime_state({"stage": stage, "status": "completed", "serverUrl": server_url})


def run_gate(*, run_dir: Path, gate: str, contract: WorkflowBundleContract) -> GateResult:
    gate_contract = contract.manifest.gates[gate]
    command = render_command(gate_contract, run_dir=run_dir)
    cwd = gate_cwd(gate=gate, contract=contract)
    log_dir = run_dir / "artifacts" / "logs"
    log_dir.mkdir(parents=True, exist_ok=True)
    log_runner_event(run_dir, gate, "running gate: " + " ".join(shlex.quote(part) for part in command))
    completed = subprocess.run(command, cwd=cwd, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=False)
    (log_dir / f"{gate}.stdout.txt").write_text(completed.stdout, encoding="utf-8")
    (log_dir / f"{gate}.stderr.txt").write_text(completed.stderr, encoding="utf-8")
    append_jsonl(
        run_dir / "artifacts" / "events.jsonl",
        {"kind": "gate", "gate": gate, "exitCode": completed.returncode, "ok": completed.returncode == 0},
    )
    return GateResult(gate, completed.returncode)


def render_command(gate: GateContract, *, run_dir: Path) -> list[str]:
    return [part.replace("{{runDirectory}}", str(run_dir)) for part in gate.command]


def gate_cwd(*, gate: str, contract: WorkflowBundleContract) -> Path:
    raw = contract.gate_cwds.get(gate)
    if raw is None:
        return contract.workflow_root
    return (contract.workflow_root / raw).resolve()


def get_or_create_session(
    *,
    contract: WorkflowBundleContract,
    store: SessionStore,
    client: OpencodeHttpClient,
    stage: str,
) -> SessionRecord:
    session_contract = contract.manifest.sessions[stage]
    existing = store.get(stage)
    if existing is not None and client.get_session(existing.session_id) is not None:
        return existing
    created = client.create_session(f"{contract.manifest.run.workflow_name}:{stage}")
    return store.upsert(stage, session_id=created.session_id, agent=session_contract.agent, prompt_count=0)


def ensure_workflow_sessions(
    *,
    run_dir: Path,
    contract: WorkflowBundleContract,
    store: SessionStore,
    client: OpencodeHttpClient,
) -> None:
    for stage in sorted(contract.manifest.sessions):
        get_or_create_session(contract=contract, store=store, client=client, stage=stage)
    refresh_attach_guide(run_dir=run_dir, server_url="<pending>", contract=contract, store=store)


def wait_for_stage_completion(
    *,
    run_dir: Path,
    stage: str,
    session_id: str,
    client: OpencodeHttpClient,
    agent_timeout_sec: float,
    contract: WorkflowBundleContract,
) -> None:
    deadline = time.monotonic() + max(agent_timeout_sec * STAGE_HARD_TIMEOUT_MULTIPLIER, OPEN_ASSISTANT_MAX_WAIT_SEC)
    previous: StageWaitObservation | None = None
    while time.monotonic() < deadline:
        observation = observe_stage_completion(run_dir=run_dir, stage=stage, session_id=session_id, client=client, contract=contract)
        if observation != previous:
            append_jsonl(
                run_dir / "artifacts" / "events.jsonl",
                {
                    "kind": "stage-observe",
                    "stage": stage,
                    "assistantStarted": observation.assistant_started,
                    "assistantCompleted": observation.assistant_completed,
                    "openAssistant": observation.open_assistant,
                    "outputsComplete": observation.outputs_complete,
                    "missingOutputs": list(observation.missing_outputs),
                },
            )
            previous = observation
        if observation.assistant_completed and observation.outputs_complete:
            return
        time.sleep(STAGE_POLL_INTERVAL_SEC)
    raise RuntimeError(f"stage {stage} did not complete before timeout")


def observe_stage_completion(
    *,
    run_dir: Path,
    stage: str,
    session_id: str,
    client: OpencodeHttpClient,
    contract: WorkflowBundleContract,
) -> StageWaitObservation:
    messages_supported = True
    assistant_started = False
    assistant_completed = False
    open_assistant = False
    message_fingerprint = None
    try:
        messages = client.list_messages(session_id)
        assistant_messages = [message for message in messages if message.get("role") == "assistant"]
        assistant_started = bool(assistant_messages)
        assistant_completed = any(_message_looks_complete(message) for message in assistant_messages)
        open_assistant = assistant_started and not assistant_completed
        message_fingerprint = _fingerprint(messages)
    except OpencodeHttpError:
        messages_supported = False
    outputs = []
    missing = []
    for path in _required_paths_for_stage(run_dir=run_dir, stage=stage, contract=contract):
        exists = path.exists()
        outputs.append((str(path), exists, path.stat().st_size if exists else None))
        if not exists:
            missing.append(str(path))
    return StageWaitObservation(
        messages_supported=messages_supported,
        assistant_started=assistant_started,
        assistant_completed=assistant_completed or not messages_supported,
        open_assistant=open_assistant,
        outputs_complete=not missing,
        message_fingerprint=message_fingerprint,
        output_fingerprint=tuple(outputs),
        missing_outputs=tuple(missing),
    )


def _message_looks_complete(message: dict[str, object]) -> bool:
    status = message.get("status")
    if isinstance(status, str) and status.lower() in {"completed", "done", "idle"}:
        return True
    if message.get("time", {}).get("completed") if isinstance(message.get("time"), dict) else False:
        return True
    return bool(message.get("parts") or message.get("content"))


def _required_paths_for_stage(*, run_dir: Path, stage: str, contract: WorkflowBundleContract) -> list[Path]:
    from prompt_assembly import required_output_paths

    return required_output_paths(
        stage,
        run_dir=run_dir,
        repo_root=REPO_ROOT,
        workflow_root=contract.workflow_root,
        manifest=contract.manifest,
    )


def print_observer_summary(
    *,
    run_dir: Path,
    server_url: str,
    contract: WorkflowBundleContract,
    store: SessionStore,
) -> None:
    refresh_attach_guide(run_dir=run_dir, server_url=server_url, contract=contract, store=store)
    guide_path = run_dir / "artifacts" / ATTACH_GUIDE_FILENAME
    print(f"[runner] runDirectory={run_dir}")
    print(f"[runner] opencodeServer={server_url}")
    print(f"[runner] attachGuide={guide_path}")
    print("[runner] attach commands:")
    print(guide_path.read_text(encoding="utf-8"))


def refresh_attach_guide(
    *,
    run_dir: Path,
    server_url: str,
    contract: WorkflowBundleContract,
    store: SessionStore,
) -> None:
    lines = [
        "# OpenCode attach commands",
        f"server={server_url}",
        f"runDirectory={run_dir}",
        "",
        "# Use /exit to leave the TUI client. It should not cancel server-owned execution.",
    ]
    for stage in sorted(contract.manifest.sessions):
        record = store.get(stage)
        if record is None:
            continue
        lines.append("")
        lines.append(f"# {stage}")
        lines.append(attach_command(server_url=server_url, session_id=record.session_id))
    path = run_dir / "artifacts" / ATTACH_GUIDE_FILENAME
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")


def attach_command(*, server_url: str, session_id: str) -> str:
    return " ".join(
        [
            "opencode",
            "attach",
            shlex.quote(server_url),
            "--dir",
            shlex.quote(str(REPO_ROOT)),
            "--session",
            shlex.quote(session_id),
        ]
    )


def log_runner_event(run_dir: Path, label: str, message: str) -> None:
    append_jsonl(run_dir / "artifacts" / "runner-events.jsonl", {"label": label, "message": message, "time": time.time()})


def append_jsonl(path: Path, payload: dict[str, object]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(payload, sort_keys=True) + "\n")


def _fingerprint(value: object) -> str:
    encoded = json.dumps(value, sort_keys=True, default=str).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()[:16]


def _stage_controls_decision(*, run_dir: Path, stage: str, contract: WorkflowBundleContract) -> bool:
    if stage != "review":
        artifact_path = run_dir / contract.manifest.decision.artifact
        return artifact_path.exists() and contract.manifest.decision.artifact in contract.manifest.stages[stage].required_outputs
    return True


def decision_matches_contract(*, run_dir: Path, contract: WorkflowBundleContract) -> bool:
    artifact_path = run_dir / contract.manifest.decision.artifact
    if not artifact_path.is_file():
        return False
    payload = json.loads(artifact_path.read_text(encoding="utf-8"))
    cursor: object = payload
    for segment in contract.manifest.decision.json_path:
        if not isinstance(cursor, dict) or segment not in cursor:
            return False
        cursor = cursor[segment]
    return cursor == contract.manifest.decision.equals


if __name__ == "__main__":
    raise SystemExit(main())
